thirdparty/cmdconfig/commands/cmdeval/cmdeval.go

```diff
@@ -176,24 +176,33 @@ func (r *EvalFnRunner) SaveFnToKptfile() {
 	}
 	if kf.Pipeline.Mutators == nil {
 		kf.Pipeline.Mutators = []kptfile.Function{}
-	} else {
-		for _, m := range kf.Pipeline.Mutators {
-			if m.Name == r.Image || m.Image == r.Image {
-				pr.Printf("skip adding image: already exists in Kptfile\n")
-				return
-			}
-			if m.Name == r.Exec || m.Exec == r.Exec {
-				pr.Printf("skip adding exec: already exists in Kptfile\n")
-				return
-			}
+	}
+	var newMutators []kptfile.Function
+	found := false
+	for _, m := range kf.Pipeline.Mutators {
+		switch {
+		case m.Image != "" && m.Image == r.Image:
+			pr.Printf("found image in Kptfile, updating...\n")
+			newMutators = append(newMutators, *r.NewFunction())
+			found = true
+		case m.Exec != "" && m.Exec == r.Exec:
+			pr.Printf("found exec in Kptfile, updating...\n")
+			newMutators = append(newMutators, *r.NewFunction())
+			found = true
+		default:
+			newMutators = append(newMutators, m)
 		}
 	}
-	kf.Pipeline.Mutators = append(kf.Pipeline.Mutators, *r.NewFunction())
-	if err = kptfileutil.WriteFile(r.Dest, kf); err != nil {
+	if !found {
+		newMutators = append(newMutators, *r.NewFunction())
+		pr.Printf("adding function to Kptfile\n")
+	}
+	kf.Pipeline.Mutators = newMutators
+	if err = kptfileutil.WriteFile(r.RunFns.Path, kf); err != nil {
```

Comment:
One thing I am not able to recall is if comments in kptfile will be preserved across these writes or not ? something to verify.

Reply:
We do not have mechanism to preserve comments in Kptfile read/write. Once the kptfile updates, comments will be removed. 
@morgante How serious it is to preserve the comments in Kptfile (I think we want to keep the license header at least)?

Reply:
Yeah it's somewhat critical that we not overwrite the Kptfile's comments.

Reply:
I see. We need to develop a new way to retrieve and update Kptfile then (e.g. using kustomize/kyaml RNode). Would blueprint KRM functions edit Kptfile as well? If so, I will put the new kptfile read/write in kpt-functions-sdk

Reply:
@droot  @mengqiy @natasha41575 I'm thinking of importing the kustomize/kyaml comments handlers to kpt to preserve the Kptfile comments. But this would introduce a non-trivial amount of code and in a RNode-like style wich is very different from kpt source code. What're your thoughts on this approach?

Reply:
@yuwenma what extra code would it introduce? Would you be able to do something here like 

```
kfNodeOriginal, _ := yaml.Parse(kf.String()) // save original kptfile content as an RNode
.... // operations to mutate kptfile
kfNode, _ := yaml.Parse(kf.String()) // save mutated kptfile content as an RNode
kyaml.CopyComments(kfNodeOriginal, kfNode) // copy the saved comments  from kfNodeOriginal to kfNode
yaml.Unmarshal(kf, kfNode.MustString()) // convert the kfNode back to kpt Kptfile struct
```

Does that make sense? Wondering if this approach would work, and if it would be relatively simple to do.

Reply:
Maybe we can introduce a wrapper around `kyaml.CopyComments` to do the conversions from kf to RNode and back?

Reply:
Another option is preserve the license header only. I'm going to merge this PR and updates the solution later on.

Reply:
> Another option is preserve the license header only. I'm going to merge this PR and updates the solution later on.

We shouldn't do that.

Would it be possible to reuse the kpt logic which automatically reattaches comments on resources after functions modify them? Ideally function authors should be able to modify the Kptfile without worrying about comment management.

Reply:
"auto reattaches comments on resources" -> this is the kustomize/kyaml RNode we discussed above.

Reply:
Tracked in https://github.com/GoogleContainerTools/kpt/issues/2952